spec/hotel_spec.rb

```diff
@@ -0,0 +1,239 @@
+      start_date = Date.new(2001, 3, 2)
+      end_date = Date.new(2001, 3, 12)
+      expect(@hotel.available_rooms(start_date, end_date)).must_equal [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 13, 14, 15, 16, 17, 18, 19]
+    end
+
+    it "desired reservation starts during and ends after existing reservation" do
+      start_date = Date.new(2001, 3, 6)
+      end_date = Date.new(2001, 3, 15)
+      expect(@hotel.available_rooms(start_date, end_date)).must_equal [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 13, 14, 15, 16, 17, 18, 19]
+    end
+
+    it "desired reservation starts before and ends after existing reservation" do
+      start_date = Date.new(2001, 3, 2)
+      end_date = Date.new(2001, 3, 15)
+      expect(@hotel.available_rooms(start_date, end_date)).must_equal [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 13, 14, 15, 16, 17, 18, 19]
+    end
+
+    it "desired reservation starts during and ends on existing reservation" do
+      start_date = Date.new(2001, 3, 6)
+      end_date = Date.new(2001, 3, 12)
+      expect(@hotel.available_rooms(start_date, end_date)).must_equal [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 13, 14, 15, 16, 17, 18, 19]
+    end
+
+    it "desired reservation begins and ends before existing reservation" do
+      start_date = Date.new(2001, 3, 2)
+      end_date = Date.new(2001, 3, 4)
+      expect(@hotel.available_rooms(start_date, end_date)).must_equal [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19]
+    end
+
+    it "desired reservation begins at end date of existing reservation and continues " do
+      start_date = Date.new(2001, 3, 12)
+      end_date = Date.new(2001, 3, 15)
+      expect(@hotel.available_rooms(start_date, end_date)).must_equal [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19]
+    end
+
+    it "desired reservations begins and ends before existing reservation" do
+      start_date = Date.new(2001, 3, 1)
+      end_date = Date.new(2001, 3, 3)
+      expect(@hotel.available_rooms(start_date, end_date)).must_equal [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19]
+    end
+
+    it "desired reservation to begin and end after existing reservation ends" do
+      start_date = Date.new(2001, 3, 13)
+      end_date = Date.new(2001, 3, 16)
+      expect(@hotel.available_rooms(start_date, end_date)).must_equal [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19]
+    end
+
+    it "existing reservation begins and ends reservation" do
+      start_date = Date.new(2001, 3, 7)
+      end_date = Date.new(2001, 3, 10)
+      expect(@hotel.available_rooms(start_date, end_date)).must_equal [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 13, 14, 15, 16, 17, 18, 19]
+    end
+
+    it "raises an Argument Error if tries to reserve room that is not available for a given day" do
+      expect { @hotel.reserve_room(Date.new(2001, 3, 5), Date.new(2001, 3, 17), 12) }.must_raise ArgumentError
+    end
+  end
```

Comment:
Aha, here is the test I asked for above. Since this exercises the `reserve_room` method, it should be grouped with those tests.